### PR TITLE
Fail when the forked test harness fails

### LINKED ISSUE #2442: sbt reports success when forked test suite fails with an exception
See https://scala-ci.typesafe.com/job/dotty-master-validate-junit/1000/console for an example:

``` scala
Internal error when running tests: java.lang.OutOfMemoryError: Java heap space
Exception in thread "Thread-7" java.io.EOFException
    at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2601)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1319)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
    at sbt.React.react(ForkTests.scala:114)
    at sbt.ForkTests$$anonfun$mainTestTask$1$Acceptor$2$.run(ForkTests.scala:74)
    at java.lang.Thread.run(Thread.java:745)
[info] Passed: Total 8, Failed 0, Errors 0, Passed 8
[success] Total time: 1721 s, completed Feb 4, 2016 8:53:52 PM
```

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #2442/#2722

Replaces #2723

### CODE REVIEW

**Comment by dwijnand** [comment]
Given this attracted some initial feedback in https://github.com/sbt/sbt/pull/2723, I'm holding back on merging this until it gets some LGTMs.

Review by @eed3si9n / @retronym plz

File: main/actions/src/main/scala/sbt/ForkTests.scala [hunk 1]
```diff
@@ -73,6 +74,8 @@ private[sbt] object ForkTests {
             os.flush()
 
             new React(is, os, log, opts.testListeners, resultsAcc).react()
+          } catch {
+            case NonFatal(e) => resultsAcc("Forked test harness failed: " + e.getMessage) = SuiteResult.Error
```
Comment (retronym) [inline main/actions/src/main/scala/sbt/ForkTests.scala]: Would still prefer:

```
scala> def throwableToString(t: Throwable): String = { import java.io._; val sw = new StringWriter; t.printStackTrace(new PrintWriter(sw)); sw.toString }
throwableToString: (t: Throwable)String

scala> println(throwableToString(new Throwable({def foo = new Throwable; foo})))
java.lang.Throwable: java.lang.Throwable
    at $line13.$read$$iw$$iw$.<init>(<console>:13)
    at $line13.$read$$iw$$iw$.<clinit>(<console>)
    at $line13.$eval$.$print$lzycompute(<console>:7)
    at $line13.$eval$.$print(<console>:6)
    at $line13.$eval.$print(<console>)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:786)
    at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1047)
    at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:638)
    at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:637)
    at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
    at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:19)
    at scala.tools.nsc.interpreter.IMain$WrappedRequest.loadAndRunReq(IMain.scala:637)
    at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:569)
    at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:565)
    at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:807)
    at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:681)
    at scala.tools.nsc.interpreter.ILoop.processLine(ILoop.scala:395)
    at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:415)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply$mcZ$sp(ILoop.scala:923)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:909)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:909)
    at scala.reflect.internal.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:97)
    at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:909)
    at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:74)
    at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:87)
    at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:98)
    at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:103)
    at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
Caused by: java.lang.Throwable
    at $line13.$read$$iw$$iw$.foo$1(<console>:13)
    ... 32 more
```

File: sbt/src/sbt-test/tests/fork-uncaught2/src/test/scala/build/MyFramework.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+class MyFramework extends sbt.testing.Framework {
+  def fingerprints = Array(new AnnotatedFingerprint { def isModule = true; def annotationName = "my" })
+  def name = "my"
+  def runner(args: Array[String], remoteArgs: Array[String], testClassLoader: ClassLoader): Runner =
+    new MyRunner(args, remoteArgs, testClassLoader)
+}
+
+class MyRunner(val args: Array[String], val remoteArgs: Array[String],
+               val testClassLoader: ClassLoader) extends sbt.testing.Runner {
+  def tasks(taskDefs: Array[TaskDef]): Array[Task] = throw new Throwable()
```
Comment (retronym) [inline sbt/src/sbt-test/tests/fork-uncaught2/src/test/scala/build/MyFramework.scala]: Could also test throwing something from the `Task`.

Comment (dwijnand) [reply sbt/src/sbt-test/tests/fork-uncaught2/src/test/scala/build/MyFramework.scala]: Thought you might be interested: throwing in Task's execute already fails the test task.

But I'm adding it to the test nonetheless.

**Comment by dwijnand** [comment]
By the way, this is green in TravisCI. So if the implementation and testing is good enough, it's ready to be merged.

**Comment by eed3si9n** [comment]
LGTM

**Comment by eed3si9n** [comment]
Forward port this please.

**Pull request merged**